vendor/github.com/tonistiigi/fsutil/send_windows.go

```diff
@@ -0,0 +1,19 @@
+package fsutil
+
+import (
+	"context"
+
+	"github.com/Microsoft/go-winio"
+)
+
+func send(ctx context.Context, s *sender) error {
+	// Windows rootfs has a few special metadata files that
+	// require extra privileges to be accessed.
+
+	// TODO: [discussion] wrapping this from the caller's side in buildkit
+	// at the last point at session/filesync/diffcopy.go:25
+	// had no effect. Suspect is to do with the goroutines.
+	return winio.RunWithPrivileges([]string{winio.SeBackupPrivilege}, func() error {

```

Comment:
> My hunch is that it's because the call breaks into goroutines and escapes the privilege binding?

I think that's indeed what's happening. The `RunWithPrivileges` function locks the current thread for just the function that is about to run, then escalates privileges. At the end it releases the thread and restores previous privileges.

If the execution happens in another thread, that thread won't have the same privileges.

One more heavy handed way to do this is to try to escalate privileges process-wide using:

https://github.com/microsoft/go-winio/blob/bdc6c112805ce13d7b235dc6257f258d958f5471/privilege.go#L108

something like

```go
winio.EnableProcessPrivileges([]string{winio.SeBackupPrivilege})
defer winio.DisableProcessPrivileges([]string{winio.SeBackupPrivilege})

// do the thing
```

Enabling process wide privileges is not ideal if it's being done in multiple places of the code base. There is a chance (I am not vastly familiar with internals) that if you revoke the privilege once you are done in a `defer`, you may remove the privilege, while it's needed by some other go routine somewhere else in the code.

There is one other place that `EnableProcessPrivilege` is used in `fsutil`, here:

https://github.com/tonistiigi/fsutil/blob/91a3fc46842c58b62dd4630b688662842364da49/copy/copy_windows.go#L62-L74

That can actually be converted to `RunWithPrivileges` as it only sets the security descriptor on a file and can run in the same thread. But my point is that enabling/disabling privileges process wide needs to be done carefully. And we need to be aware of the potential hard to debug scenarios when something fails intermittently  without obvious reason.

Adding this to `fsutil` might make sense, but it feels equally heavy handed to require elevated privileges for every `Walk()`, even when it's called for paths that we should have access to normally. The `fsutil` package is generic. The case we're trying to address here is specific to the particular need we have here.

@tonistiigi what are your thoughts here?

Reply:
Thanks for that. Sure, I agree we need caution for process-wide escalation.

I had attempted an option where I retry running privileges only after a `os.IsPermission(err)` but that become too late since the context is cancelled immediately there's the first error. Is this close to a possible option?